src/main/java/com/rockset/jdbc/RocksetResultSet.java

```diff
@@ -1567,55 +1573,49 @@ private static void log(String msg) {
-          while (fields.hasNext()) {
-            Map.Entry<String, JsonNode> field = fields.next();
-            String fieldName = field.getKey();
-            if (fieldNames.contains(fieldName)) {
-              // This fieldname was already found to have a non-null type
-              // in a previous record.
-              continue;
-            }
-            JsonNode value = field.getValue();
-            Column.ColumnTypes type = Column.ColumnTypes.fromValue(value.getNodeType().toString());
-            // Skip over the fields with null type unless all values for that
-            // field are null
-            if (type.equals(Column.ColumnTypes.NULL) && i != response.getResults().size() - 1) {
-              continue;
-            }
-            if (type.equals(Column.ColumnTypes.STRING)) {
-              java.time.format.DateTimeFormatter format = TIMESTAMP_PARSE_FORMAT;
-              try {
-                LocalDateTime.parse(value.asText(), format);
-                type = Column.ColumnTypes.TIMESTAMP;
-              } catch (DateTimeParseException e) {
-                // ignore
-              }
+          JsonNode docRootNode = OBJECT_MAPPER.readTree(OBJECT_MAPPER.writeValueAsString(onedoc));
+
+            for (Iterator<Map.Entry<String, JsonNode>> it = docRootNode.fields(); it.hasNext(); ) {
+                Map.Entry<String, JsonNode> field = it.next();
+                String fieldName = field.getKey();
+                if (fieldTypes.containsKey(fieldName) && !fieldTypes.get(fieldName).equals(Column.ColumnTypes.NULL)) {
+                  // This fieldname was already found to have a non-null type
+                  // in a previous record.
+                  continue;
+                }
+                JsonNode value = field.getValue();
+                Column.ColumnTypes type = valueToColumnType(value);
+                log("getColumns::column name " + fieldName + " type: " + type.toString());
+                fieldTypes.put(fieldName, type);
             }
-            if (type.equals(Column.ColumnTypes.OBJECT)) {
-              if (value.get("__rockset_type") != null) {
-                type = Column.ColumnTypes.fromValue(value.get("__rockset_type").asText());
-              }
-            }
-            log("getColumns::column name " + fieldName + " type: " + type.toString());
-            Column c = new Column(fieldName, type);
-            out.add(c);
-            fieldNames.add(fieldName);
-          }
         }
-      } else if (response.getColumnFields() != null && response.getColumnFields().size() > 0) {
+
+        // If we know the desired field ordering through explicit projection, use that ordering.
+        // Otherwise, just iterate over in arbitrary ordering
+        Collection<String> fields = orderedFields.isEmpty() ? fieldTypes.keySet() : orderedFields;
+        for(String field : fields){
+          out.add(new Column(field, fieldTypes.get(field)));
```

Comment:
If a field is null in all rows, then it wouldn't be in `fieldTypes` as they get skipped on line 1599 (previously this was handled on like 1595). Idk if `new Column(field, null)` will interpret it as NULL type - if not this needs to be handled

Reply:
This is handled by the column conversion, I added a test case to make sure

Reply:
what do you mean by "column conversion"?

Reply:
https://github.com/rockset/rockset-java-client/blob/dac440d557c24025b6b69add829feaca82029b96/src/main/java/com/rockset/jdbc/RocksetResultSet.java#L1641

Reply:
Thanks! I figured the column conversion part but I incorrectly understood that `fieldTypes` wouldn't have an entry for the field until the last row based on the previous implementation which is now being handled differently by tracking the NULL type fields as we see and overriding them when we encounter a non-null type.